src/catalog/src/persist.rs

```diff
@@ -0,0 +1,1084 @@
+                    }
+                    StateUpdateKind::DefaultPrivilege(key, value) => {
+                        apply(&mut self.snapshot.default_privileges, key, value, diff);
+                    }
+                    StateUpdateKind::Epoch(epoch) => {
+                        return Err(DurableCatalogError::Fence(format!(
+                            "current catalog epoch {} fenced by new catalog epoch {}",
+                            self.epoch, epoch
+                        )));
+                    }
+                    StateUpdateKind::IdAllocator(key, value) => {
+                        apply(&mut self.snapshot.id_allocator, key, value, diff);
+                    }
+                    StateUpdateKind::IntrospectionSourceIndex(key, value) => {
+                        apply(&mut self.snapshot.introspection_sources, key, value, diff);
+                    }
+                    StateUpdateKind::Item(key, value) => {
+                        apply(&mut self.snapshot.items, key, value, diff);
+                    }
+                    StateUpdateKind::Role(key, value) => {
+                        apply(&mut self.snapshot.roles, key, value, diff);
+                    }
+                    StateUpdateKind::Schema(key, value) => {
+                        apply(&mut self.snapshot.schemas, key, value, diff);
+                    }
+                    StateUpdateKind::Setting(key, value) => {
+                        apply(&mut self.snapshot.settings, key, value, diff);
+                    }
+                    StateUpdateKind::StorageUsage(_, _) => {
+                        // We can ignore storage usage since it's not cached in memory.
+                    }
+                    StateUpdateKind::SystemConfiguration(key, value) => {
+                        apply(&mut self.snapshot.system_configurations, key, value, diff);
+                    }
+                    StateUpdateKind::SystemObjectMapping(key, value) => {
+                        apply(&mut self.snapshot.system_object_mappings, key, value, diff);
+                    }
+                    StateUpdateKind::SystemPrivilege(key, value) => {
+                        apply(&mut self.snapshot.system_privileges, key, value, diff);
+                    }
+                    StateUpdateKind::Timestamp(key, value) => {
+                        apply(&mut self.snapshot.timestamps, key, value, diff);
+                    }
+                },
+                invalid_update => {
+                    panic!("invalid update in consolidated trace: {:?}", invalid_update);
+                }
+            }
+        }
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl ReadOnlyDurableCatalogState for PersistCatalogState {
+    fn epoch(&mut self) -> Epoch {
+        self.epoch
+    }
+
+    async fn get_catalog_content_version(&mut self) -> Result<Option<String>, CatalogError> {
+        self.confirm_leadership().await?;
```

Comment:
What do you think about adding a new method on `PersistCatalogState` that is:
```
async fn with_snapshot<'a, F, T>(&'a mut self, f: F) -> Result<T, CatalogError> 
where
  F: FnOnce(&'a mut Snapshot) -> Result<T, CatalogError>,
{
  self.confirm_leadership().await?;
  f(self.snapshot)
}
```
This way all of these methods could call `.with_snapshot(...)` and be less likely to forget to confirm leadership?

It's not a huge difference, so I'll leave it up to you if you want to make this change

Reply:
Great idea! Updated.